### PR TITLE
Allow * in variable string literal defaults

### LINKED ISSUE #4959: Asterisk in default variable value leads to value interpreted as string and not interpolated
```allowedOrigin: ${opt:allowedOrigin, 'beta.blabla.net'}```
works fine
```allowedOrigin: ${opt:allowedOrigin, '*.blabla.net'}``` 
 doesnt work

in second case allowedOrigin will be interpreted as a string e.g. its value will be ${opt:allowedOrigin, '.blabla.net'} instead of *.blabla.net

### LINKED ISSUE #5225: Cannot pass * as default value
# This is a Bug Report

## Description

* What went wrong?

When I set `${env:ALLOW_ORIGIN, "*"}` it gets passed as is.

* What did you expect should have happened?

As `${env:ALLOW_ORIGIN, "foo"}` works, I'd expect `${env:ALLOW_ORIGIN, "*"}` to work also.

* What was the config you used?

```
service: community-slack

provider:
  name: aws
  runtime: nodejs6.10
  environment:
    SLACK_TEAM: ${env:SLACK_TEAM}
    SLACK_TOKEN: ${env:SLACK_TOKEN}
    ALLOW_ORIGIN: ${env:ALLOW_ORIGIN, "*"}

functions:
  doInvite:
    handler: handler.doInvite
    events:
      - http:
          path: invite
          method: post
          cors:
            origin: ${env:ALLOW_ORIGIN, "*"}
            headers:
              - Content-Type
              - X-Amz-Date
              - Authorization
              - X-Api-Key
              - X-Amz-Security-Token
              - X-Amz-User-Agent
            allowCredentials: false
plugins:
  - serverless-stack-output

custom:
  output:
    file: .serverless/output-${opt:stage, self:provider.stage}.json

package:
  exclude:
    - 'node_modules/**'
    - 'docs/**'
    - '*.md'
    - 'LICENSE'
    - '*.sh'
    - '*.env'
    - 'package-lock.json'
    - 'package.json'
    - '.vscode/**'
```

* What stacktrace or error message from your provider did you see?

No error.

## Additional Data

* ***Serverless Framework Version you're using***: 1.30.1
* ***Operating System***: macOS

### PR DESCRIPTION
**Pull request by dschep**
## What did you implement:
closes #5225
closes #4959
closes #4967 (older PR) - this implementation isn't quite the same. It just adds `*` to the regex. Quotes are required for it to work, when quotes are omitted, nothing resolves(see tests).

## How did you implement it:
Added `*` to the variable syntax regex

## How can we verify it:
Create a service with `${opt:foobar, "*"}` in a variable and run `sls print` to check that the variable is correctly substituted.

## Todos:

- [X] Write tests
- [X] Fix linting errors
- [X] Make sure code coverage hasn't dropped
- [X] Provide verification config / commands / resources
- [X] Enable "Allow edits from maintainers" for this PR
- [X] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by dschep** [comment]
cc @beeceej & @exoego

File: lib/classes/Service.js [hunk 1]
```diff
@@ -19,7 +19,7 @@ class Service {
     this.provider = {
       stage: 'dev',
       region: 'us-east-1',
-      variableSyntax: '\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}',
+      variableSyntax: '\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)*]+?)}',
```
Comment (erikerikson) [inline lib/classes/Service.js]: As per comment on 11-13, see https://github.com/serverless/serverless/blob/5c2c1ae405514b5c7bb61493d9be9f7dd62e9252/lib/plugins/print/print.js#L57

File: lib/classes/Service.test.js [hunk 2]
```diff
@@ -31,7 +31,7 @@ describe('Service', () => {
       expect(serviceInstance.provider).to.deep.equal({
         stage: 'dev',
         region: 'us-east-1',
-        variableSyntax: '\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)]+?)}',
+        variableSyntax: '\\${([ ~:a-zA-Z0-9._\'",\\-\\/\\(\\)*]+?)}',
```
Comment (erikerikson) [inline lib/classes/Service.test.js]: see also print tests: https://github.com/serverless/serverless/blob/5c2c1ae405514b5c7bb61493d9be9f7dd62e9252/lib/plugins/print/print.test.js#L35

**Comment by beeceej** [comment]
@dschep I am fine with you owning this issue from here on out, especially since you may be a bit closer to the issue at hand.

**Comment by dschep** [comment]
Thanks for the notes @erikerikson!

I've updated the print tests.

re: the previous PR, I've looked at it (it's marked in the description to close that). I'm not sure what the purpose of the lookahead assertion is (i'd love son insight on that @beeceej). Also, it seems to introduce the ability to use unquoted default literals, which I don't like at all and seem more likely to introduce a regression for some one. And speaking to your concern over regressions, I'm pretty hopeful this won't cause issues, mostly because before the change, their variable references were resolving to the literal variable ref (which would also contain a `*`) which seems just as likely to cause problems.

**Comment by erikerikson** [comment]
@dschep I think you're right that the risk of regression is low here but I was leaving the heavy thinking to you.  There was another PR where someone had been arguing for allowing `*`s in the file variable type to support paths including `*` where there was discussion.  Probably not too important to dig that up.

Regarding your test "should do nothing useful", I'm not sure what "useful" means there.   What would expect to result to be if you defined `custom.*` to a value like `'foo'`?  It appears to me that it attempted to resolve `*` as an attribute of the root property `custom` and ended up resolving it to `undefined` which seems potentially useful to me.  To avoid this you would need to only accept quoted `*`, but doing that is not trivial (it might be possible with the new regex capabilities but they're not assumable yet unless this is being babeled) - you can see the implementation of detecting unquoted verses quoted comas in Variables.js as part of the override statement detection and splitting logic.

**Comment by dschep** [comment]
Ah. Good points @erikerikson. I think this is a risk we have to accept IMO. Can't support every odd workflow out there 😉

**Comment by beeceej** [comment]
@erikerikson @dschep  In reference to only resolving quoted `*'s`, if I remember correctly this is what my look behind was doing. It's a bit verbose because of lacking support for labeled groups.

for reference this is what it looks like, 

```
\${(([ ~:a-zA-Z0-9._'",\\-\\/\\(\\)]|\*(?=([ ~:a-zA-Z0-9._'",\\-\\/\\(\\)]+?)('|")))+?)}
```

not saying this correct or well tested (I'm not the best regexer) but may provide a decent place to start if we wish to only allow quoted `*'s`

**Comment by erikerikson** [comment]
Cool.  Just trying to support intentional decision making and clarify scope.

Thoughts on my question?
> What would expect to result to be if you defined `custom.*` to a value like `'foo'`?

It's an uncovered case and would clarify your intention.

**Comment by erikerikson** [comment]
@dschep resolving the variable to undefined is doing something.  Thus my question.

FWIW, my last crossed on the wires with the previous three.

**Comment by pmuens** [comment]
Also cross-referencing https://github.com/serverless/serverless/pull/5211 here where smth. like this was done before.

**Comment by dschep** [comment]
@pmuens, the only open quesiton is if a user includes a `*` in a variable reference (not the default), it' currently ends up resolving to undefined. Is that OK, or should it through an error?

**Comment by pmuens** [comment]
@d2s thanks for the update 👍

> @pmuens, the only open quesiton is if a user includes a * in a variable reference (not the default), it' currently ends up resolving to undefined. Is that OK, or should it through an error?

🤔 good question. Personally I'd prefer if we throw an error saying that this is not supported ATM because I dealt with some `undefined` sneaking into resolved variable strings and than e.g. had a stage `undefined` which was annoying (To be fair it was in the components project).

The only last question I have is if it's a breaking change if we throw an error. As I understand it this is an addition and we're not altering anything here, correct?

**Comment by pmuens** [comment]
Any updates on this @dschep?

Can I help you with anything here?

**Pull request merged**